fleet/ssh.go

```diff
@@ -2,49 +2,158 @@ package fleet
 
 import (
 	"fmt"
+	"io"
 	"net"
-	"os"
+	"net/http"
+	"net/url"
+	"sync"
 	"time"
 
 	"github.com/coreos/fleet/ssh"
+	"github.com/giantswarm/inago/logging"
 )
 
-// SSHTunnel contains the information needed to create a new ssh client
-type SSHTunnel struct {
+// SSHTunnelConfig contains the information needed to create a new ssh tunnel.
+type SSHTunnelConfig struct {
+	Endpoint              url.URL
+	KnownHostsFile        string
+	Logger                logging.Logger
+	StrictHostKeyChecking bool
 	Tunnel                string
-	Username              string
 	Timeout               time.Duration
-	StrictHostKeyChecking bool
-	KnownHostsFile        string
+	Username              string
+}
+
+func DefaultSSHTunnelConfig() SSHTunnelConfig {
+	URL, err := url.Parse("file:///var/run/fleet.sock")
+	if err != nil {
+		panic(err)
+	}
+
+	newConfig := SSHTunnelConfig{
+		Endpoint:              *URL,
+		KnownHostsFile:        "~/.fleetctl/known_hosts",
+		Logger:                logging.NewLogger(logging.DefaultConfig()),
+		StrictHostKeyChecking: true,
+		Tunnel:                "",
+		Timeout:               10 * time.Second,
+		Username:              "core",
+	}
+
+	return newConfig
+}
+
+// SSHTunnel provides a synchronized wrapper around http.RoundTripper for safe
+// use of tunneled ssh connections.
+type SSHTunnel interface {
+	http.RoundTripper
+
+	// IsActive determines whether the tunnel should be used. This is decided by
+	// the given tunnel configuration. In case this is empty, the SSH tunnel is
+	// not active.
+	IsActive() bool
 }
 
-// NewDialFunc returns an http.Dial function which uses the given ssh tunnel as
-// to connect to the endpoint
-func (t *SSHTunnel) NewDialFunc(config Config) func(string, string) (net.Conn, error) {
-	sshClient, err := ssh.NewSSHClient(t.Username, t.Tunnel, t.getChecker(), true, t.Timeout)
+// NewSSHTunnel creates a new SSH tunnel that is configured with the given
+// settings.
+func NewSSHTunnel(config SSHTunnelConfig) (SSHTunnel, error) {
+	newSSHTunnel := &sshTunnel{
+		SSHTunnelConfig: config,
+		Mutex:           sync.Mutex{},
+	}
+
+	newDialFunc, err := newSSHTunnel.NewDialFunc()
 	if err != nil {
-		fmt.Printf("failed to initialize ssh client: %v\n", maskAny(err))
-		os.Exit(1)
+		return nil, maskAny(err)
+	}
+	newSSHTunnel.HTTPTransport = &http.Transport{
+		Dial: newDialFunc,
 	}
-	tunnelFunc := sshClient.Dial
 
-	if config.Endpoint.Scheme == "unix" || config.Endpoint.Scheme == "file" {
-		target := config.Endpoint.Path
-		tunnelFunc = func(string, string) (net.Conn, error) {
-			cmd := fmt.Sprintf(`fleetctl fd-forward %s`, target)
+	return newSSHTunnel, nil
+}
+
+type sshTunnel struct {
+	SSHTunnelConfig
+
+	HTTPTransport http.RoundTripper
+	Mutex         sync.Mutex
+}
+
+func (t *sshTunnel) IsActive() bool {
+	return t.Tunnel != ""
+}
+
+// NewDialFunc returns an http.Dial function which uses the given ssh tunnel to
+// to connect to the configured endpoint.
```

Comment:
Done

On Fri, Apr 8, 2016 at 4:58 PM, Joseph Salisbury <notifications@github.com>
wrote:

> In fleet/ssh.go
> <https://github.com/giantswarm/inago/pull/204#discussion_r59037731>:
>
> > +	return newSSHTunnel, nil
> > +}
> > +
> > +type sshTunnel struct {
> > +	SSHTunnelConfig
> > +
> > +	HTTPTransport http.RoundTripper
> > +	Mutex         sync.Mutex
> > +}
> > +
> > +func (t *sshTunnel) IsActive() bool {
> > +	return t.Tunnel != ""
> > +}
> > +
> > +// NewDialFunc returns an http.Dial function which uses the given ssh tunnel to
> > +// to connect to the configured endpoint.
>
> Tiny comment, but you have "to" here twice.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/giantswarm/inago/pull/204/files/7c644b9fba7cc6e6ad25be21356519896ffb6008#r59037731>
>



-- 
--
Tim Schindler
Developer @ Giant Swarm GmbH
tim@giantswarm.io
Web: http://giantswarm.io
Twitter: http://twitter.com/giantswarm
Handelsregister: Amtsgericht Köln, HRB 81484 · Sitz der Gesellschaft: Köln
· Geschäftsführer: Timo Derstappen, Henning Lange, Oliver Thylmann
The latest from Giant Swarm (@giantswarm). Simple Service Orchestration,
built for developers. Cologne, Germany